include/small/rlist.h

```diff
@@ -290,6 +291,12 @@ rlist_cut_before(struct rlist *head1, struct rlist *head2, struct rlist *item)
 #define rlist_first_entry(head, type, member)				\
 	rlist_entry(rlist_first(head), type, member)
 
+/**
+ * return first entry for nonempty list and NULL otherwise
+ */
+#define rlist_first_entry_safe(head, type, member)			       \
```

Comment:
It looks like the problem is actually in our implementation of `offsetof`:

https://github.com/tarantool/small/blob/cb4b61851b9b08f713b33eea18ed6680dd3bbfbe/include/small/rlist.h#L46

Try replacing it with `__builtin_offsetof`:

```diff
diff --git a/include/small/rlist.h b/include/small/rlist.h
index f4d905ad819c..3cd3eeabb4df 100644
--- a/include/small/rlist.h
+++ b/include/small/rlist.h
@@ -280,8 +280,7 @@ rlist_cut_before(struct rlist *head1, struct rlist *head2, struct rlist *item)
  * return entry by list item
  */
 #define rlist_entry(item, type, member) ({				\
-	const typeof( ((type *)0)->member ) *__mptr = (item);		\
-	(type *)( (char *)__mptr - offsetof(type,member));		\
+	(type *)( (char *)(item) - __builtin_offsetof(type,member));	\
 })
 
 /**
@@ -378,7 +377,7 @@ delete from one list and add_tail as another's head
  * points to the list head.
  */
 #define rlist_entry_is_head(item, head, member)				\
-	((char *)(item) + offsetof(typeof(*item), member) == (char *)(head))
+	((char *)(item) + __builtin_offsetof(typeof(*item), member) == (char *)(head))
 
 /**
  * foreach through all list entries
```

Note:
 - We should use the builtin only if it's available (see `__has_builtin`).
 - There a few more places where we define `offsetof`. We need to update them all.

Reply:
Fixed, i guess

Reply:
1. Please update the commit message and the PR description.
2. Please create a PR for the Tarantool repo that would do the following (all in separate commits):
   - Bump the small library to this commit.
   - Fix all `offsetof` macros in Tarantool.
   - Enable the UBSan checks triggering the failure, as described in https://github.com/tarantool/tarantool/issues/10143#issuecomment-2222668971.
3. Please ensure that the PR for the Tarantool repo doesn't fail on CI. You'll need to enable ASAN workflows by setting the `asan-ci` label there.
4. Please replace `Fixes https://github.com/tarantool/tarantool/issues/10143` with `Needed for https://github.com/tarantool/tarantool/issues/10143` because the issue will actually be fixed by the PR for the Tarantool repo.